### PR TITLE
Subscription Block: Pass form ID as `redirect_fragment` instead of enqueueing `jetpack-subscriptions-js`

### PR DESCRIPTION
**Pull request by westonruter**
This is a follow-up on #18015.

I found that #20763 was causing AMP validation errors due to the enqueueing of a custom script on a page. The introduced script scrolls the subscriptions widget into view if a certain DOM element was on the page. Beyond not working in AMP due to the custom script, this is not ideal because adding a new external script for every page load negatively impacts frontend performance. It also doesn't work if JS is turned off.

I believe the the better fix for #18015 is merely to make sure that the ID for the `form` is passed as the `redirect_fragment` input value. Previously it was relying on the widget wrapper element to have an ID, but this appears to be no longer in place as of WordPress 5.8.

#### Changes proposed in this Pull Request:

* Revert #20763 to instead scroll to subscription form by passing the form ID as the `redirect_fragment`.

#### Does this pull request change what data or activity we track or use?

No.

#### Testing instructions:

1. Add a Subscription form to a page.
2. Submit the form (on an AMP page and a non-AMP page)
3. When the subsequent page loads, ensure you are scrolled back to the form you submitted.

### CODE REVIEW

**Comment by jeherve** [comment]
@andres-blanco Do you think you could take a look?

Thanks!

**Comment by kwight** [comment]
Hm, I don't see this working because of a mismatch between the ID in the URL and the form IDs (testing the branch at https://gradual-hoverfly.jurassic.ninja/subscribe/).

Screen Shot 2021-09-15 at 12 04 16 PM

@andres-blanco do you see the same thing? @westonruter any ideas? And thanks for the PR!

**Comment by westonruter** [comment]
Humm. I'm not seeing that issue. I've deployed the changes to my dev site. Compare:

Non-AMP: https://dev-westonruter.pantheonsite.io/2018/02/21/spoken-word-read-along-tts/#subscribe-field-blog_subscription-2
AMP: https://dev-westonruter.pantheonsite.io/2021/01/20/speaking-foreign-language/#subscribe-field-blog_subscription-2

https://user-images.githubusercontent.com/134745/133516064-195cb912-6466-4de6-ae62-e6a2ef1b2f4f.mov

https://user-images.githubusercontent.com/134745/133516078-47662b58-8ca4-4fd7-8a88-cef07c1049cb.mov

I'm getting redirected to `#blog_subscription-2` in both cases. For you, however, the ID base is `#subscribe-blog-` followed by different numbers for each request. The difference is that you're using the widget in post content whereas I'm using it in a classic widget sidebar. I can reproduce the problem when the widget is in post content. The issue comes down to this:

https://github.com/Automattic/jetpack/blob/1759f2a901d744d3ee78ff02bde9699110952fe5/projects/plugins/jetpack/modules/subscriptions/views.php#L264

The `wp_rand( 450, 550 )` is the problem. That doesn't seem ideal regardless. It seems it should be an auto-incremented ID `$instance_count` in the same way as is being used for the WPCOM version of the form:

https://github.com/Automattic/jetpack/blob/1759f2a901d744d3ee78ff02bde9699110952fe5/projects/plugins/jetpack/modules/subscriptions/views.php#L278-L280

I think this change would do the trick:

```diff
- $widget_id                    = esc_attr( ! empty( $args['widget_id'] ) ? esc_attr( $args['widget_id'] ) : wp_rand( 450, 550 ) );
+ $widget_id                    = esc_attr( ! empty( $args['widget_id'] ) ? esc_attr( $args['widget_id'] ) : self::$instance_count;
```

I'm not sure why a random number was being used in the first place.

**Comment by westonruter** [comment]
I've implemented that in 7ac91ba, as well as clean up `esc_attr()` usage. (Apologies for the generic commit message.)

**Comment by westonruter** [comment]
Aside: The AMP experience for this form could be greatly improved if the endpoint detected an AMP submission (via `Accept: application/json` request header) and return a JSON object response a `message` key, instead of redirecting to a fragment on the same page. For example, at the end of `\Jetpack_Subscriptions::widget_submit()` before the existing `wp_safe_redirect( "$redirect#$redirect_fragment" )` call:

```php
if ( wp_is_json_request() ) {
	wp_send_json(
		array(
			'message' => $result, // @todo Obtain the actual message as computed in Jetpack_Subscriptions_Widget.
		),
		'success' === $result ? 200 : 400
	);
}
```

That provides for a much better user experience.

**Comment by andres-blanco** [comment]
@westonruter thanks for the contribution! there's a small problem on the wpcom side since the form is not rendered at all when a message is displayed.

I will add a small fix to this PR tomorrow.

**Comment by andres-blanco** [comment]
* Added a fix for this to work on wpcom
* Brought back the js file since it might break automation on wpcom side

File: projects/plugins/jetpack/modules/subscriptions/views.php [hunk 1]
```diff
@@ -246,14 +238,24 @@ static function render_widget_status_messages( $instance ) {
...
 				esc_attr( $border_color ),
 				wp_kses_post( $message )
 			);
 		}
 	}
 
+	/**
+	 * Generates the redirect fragment used after form submission.
+	 */
+	protected static function get_wpcom_redirect_fragment() {
```
Comment (kwight) [inline projects/plugins/jetpack/modules/subscriptions/views.php:255]: We looked at this together today; @andres-blanco should be able to do an update that has no WP.com references, works the same on both JP and WP.com, and doesn't require the script at all in the PR.

**Review by jeherve — CHANGES_REQUESTED** [review]
Could you address the linting errors? Even though they were not introduced in this PR, they'll stop up from merging this since those lines were modified in this PR.

Thank you!

**Review by kwight — DISMISSED** [review]
This and D67171-code work well for me. Thanks everyone!

**Comment by andres-blanco** [comment]
I've rebased to latest master

**Review by jeherve — APPROVED** [review]

**Comment by andres-blanco** [comment]
wpcom changeset: 232609.

**Pull request merged**